### ISSUE TITLE
Training on piano dataset, loss floats around 2.5

### ISSUE THREAD

**Issue by robinsloan** [richardassar/SampleRNN_torch#1]
Hi there!

First of all: thanks for a great code release! This project is incredibly well-packaged and easy to get started with.

Using your pretrained model, I was able to generate some really nice piano samples. However, when I retrieve the piano dataset and attempt to train a model myself, I consistently get a run that looks like this --

![image](https://cloud.githubusercontent.com/assets/40284/24919224/d2cf26d4-1e97-11e7-8d29-d015f78a359c.png)

-- with commensurately bad/noisy samples.

I'm training with all the default settings.

Any ideas or tips?

(One miscellaneous thing: I'm running on a single GPU, but `train.lua` remains flagged for multi-GPU. If the multi-GPU flag is set to `false`, the code bombs out with an error. I think I can see where to fix it, but haven't tried to do so yet 😅)

**Comment by richardassar** [comment]
Hey, thanks for the feedback. I really appreciate it :)

Also thanks for the bug report, let's work through it.

Have you pulled recently? I had originally published the code with cudnn rnn
enabled by default, this was a huge mistake because lack of weight
normalization in the GRU seems to causes models to blow up when they have
converged. In your case it seems to be blowing up after very few
mini-batches though, which is very suspicious.

Make sure you are starting a new session, not resuming the one you
downloaded. If you're training with a different minibatch size or TBPTT
step size (which I assume you are to fit within GPU memory limits) then the
optim_state would not be valid. Please try a pull, delete that session and
start a new one.

As for the arguments, this is something I noticed the other day also.
Having args which default to "true" seems problematic with torch cmdline.
The only workaround is to negate the logic, so have "-no_multigpu" to
disable multigpu support. I don't like this at all so I can probably just
parse the "arg" parameter myself to detect when they're explicitly being
set to false, which would fix your issue. Alternatively I could default all
current boolean args to false and force the user to specify them, could add
these as session parameters so you don't have to keep specifying them (but
then overriding them to false would require custom arg parse code).

tl;dr - Bool arg parsing It's a mess. I'll address this but for now you can
hack it in by setting
https://github.com/richardassar/SampleRNN_torch/blob/master/train.lua#L119
to false.

**Comment by robinsloan** [comment]
Hey @richardassar! Thanks very much for the quick and helpful reply. Having just synced my repo to the latest, I'm training now -- a fresh session -- and I will report back!

On the multi-GPU front: after setting the flag to `false`, `train.lua` fails at line 395, `net:syncParameters`. I think I can probably figure this out and submit a patch… it might be as simple as hiding those `syncParameters` calls behind a multi-GPU flag check?

**Comment by richardassar** [comment]
Hey, ah yes you're right. Nicely caught.

I've just pushed a fix for that.

**Comment by robinsloan** [comment]
Here's a plot after training overnight:

![image](https://cloud.githubusercontent.com/assets/40284/24965104/f32b69c2-1f57-11e7-9165-1cbc5fd9516e.png)

Same basic results 😓

I'll pull the very newest code and try again. In the meantime, any conjectures are most welcome!

**Comment by richardassar** [comment]
Does this happen right after sampling after the first epoch?

If so then it's most certainly to do with a bug I fixed in the torch nn
package, concerning the nn.Contiguous module.

Please make sure you have this commit
https://github.com/torch/nn/commit/15977951e84621d4109ec0f4b724eff57486e5db

**Comment by richardassar** [comment]
Yeah, looks like you're on minibatch 33, so I think it's highly likely you
need to upgrade the nn package.

I will put a note about this in README.md

**Comment by robinsloan** [comment]
Yesss:

![image](https://cloud.githubusercontent.com/assets/40284/24992757/ec38ee84-1fd7-11e7-8580-2f32cbd22dff.png)

Thanks for the assistance, Richard -- I really appreciate it. I'll share some samples after training on a new/different dataset!

**Comment by richardassar** [comment]
Awesome, glad we got it sorted.

Have fun.

**Comment by greaber** [comment]
@richardassar, I was intrigued by your initial guess that the problem might have bin that @robinsloan was using an old version of the code that didn't have weight norm.  You wrote, "I had originally published the code with cudnn rnn enabled by default, this was a huge mistake because lack of weight normalization in the GRU seems to causes models to blow up when they have converged."

In the paper, they say they use weight norm in all the linear layers except the embedding layer "to accelerate training."  First of all, I am confused about what they mean by a "linear layer".  I would normally understand a linear layer to be one that does not use a (non-linear) activation function, but the reference code, if I read it correctly, seems to apply weight norm even within RNN where non-linear activation functions are being used (so you seem to be doing the right thing in your port, but I am still a bit confused about what the paper is saying and whether there are any layers to which they do not apply weight norm because they are "non-linear" layers).

More importantly, do you have any idea why the lack of weight normalization would cause the models to blow up after converging?  I would like to understand this phenomenon.  Did you test if the reference code exhibits the same problem with weight norm turned off?  The paper, after all, just says that weight norm speeds up training.

**Comment by richardassar** [comment]
Hi @greaber thanks for the questions, you raise some important points.

> First of all, I am confused about what they mean by a "linear layer". I would normally understand a linear layer to be one that does not use a (non-linear) activation function ...

Correct. This means any linear projection in the model that takes vectors from one dimensionality to another. These are disguished from non-linearities / activation functions or convolutions, which do not come under this definition. 

The RNNs (GRU or LSTM) contain linear projections to map from inputs and hidden states to hidden activations and outputs. These are unrolled in time.

It comes down to definitions, some people think of layers as including the non-linearities, fitting in with the biological interpretation of layers of neurons, but we can think of layers instead as discrete steps in our model.

The embedding, although linear, is not weight-normalized and probably doesn't need to be.

> More importantly, do you have any idea why the lack of weight normalization would cause the models to blow up after converging?

Yes, I have some ideas. I'll explain below.

> I would like to understand this phenomenon. Did you test if the reference code exhibits the same problem with weight norm turned off?

Embarrassingly, no, and I definitely should have. I am confident that my implementation (except the learnable h0 state, which is not critical) is nearly identical to the Theano implementation. I've a strong hunch that the authors encountered this problem when developing the model and found weight-normalization helpful. Layer Normalization is also likely to help.

If you can then please run the reference implementation yourself with weight_norm disabled and I'm sure you will see the same issue. You should also print out the gradient norm, you'll see the gradient explode.

Why does weight normalization help stabilise this form of training?

During a single TBPTT update you are accumulating the gradient over some number of timesteps, if the model finds itself falling into a local optima the trajectory of the optimisation can get lost, more catastrophically with per-parameter adaptive optimisers such as Adam, causing future updates to generate larger gradient signal which gradually accumulate until a single huge gradient spike causes the network to end up jumping into a region of parameter space from which it can never escape. In this case the network is dead, training has been derailed, likely due to a small sub-portion of the network being corrupted as the update was not spread out over a larger number of parameters.

Weight normalization decomposes the linear projections into a vector lying on the unit sphere with a single scaling factor (per output neuron). This reparameterisation means we can explore the scale and direction of each basis vector, adding useful degrees of freedom without introducing underdetermination to the model, avoiding saddle points and allowing the minibatches with a high degree of variance. Weight normalization will alter the loss surface in a beneficial manner by spreading out these pathological areas into the extra dimensions.

I believe the problem would be less prevalent with continuous output, instead of a categorical distribution over amplitudes, however in the paper they say worse results were observed using this approach.

Weight normalization does *much* more than speed up training, it adds stability under high variance in the training data and increases the model's capacity to generalise.

**Comment by greaber** [comment]
OK, I am running the reference code now, and I will let you know how it goes.

So your guess is that there is a single bad update step that kills the network, and that the reason it is possible to have such a step is that with TBPTT you can get larger steps (even though we clip the gradient)?  And using an optimizer like Adam or rmsprop that tries to do bigger updates of rarely updated parameters can make it worse because it might be that it is precisely seldom-updated parameters that cause the problem? However, there is no numerical problem, it is just that we are getting into a pathological region of parameter space?  I guess this sounds plausible, but I'm not sure if it is true.

I have experienced a similar phenomenon (but possibly it had a different origin) when training a Wavenet, so I wonder if TBPTT is critical or not.  In the wavenet case, at least sometimes, reducing the learning rate was enough to fix it.  I never tried weight norm with the wavenet, but it could be a good idea.

**Comment by richardassar** [comment]
At first I considered numerical stability of softmax vs logsoftmax, or
tanh/sigmoid approximations used within CUDNN. I ruled these hypotheses out
however.

**Comment by greaber** [comment]
Thanks for your feedback.  FWIW, I tried running the reference code on the "MUSIC" dataset, and both with and without weight norm it crashed after 6000 iterations with an IndexError.  I didn't use the exact versions of all the libraries that I was supposed to, so this could be the problem.  Perhaps I will try to figure out what went wrong and try again, but I am not sure if it is a good use of my time to debug this, so I might not.

```
[...]
4000 4500 5000 5500 6000
Traceback (most recent call last):
  File "models/three_tier/three_tier.py", line 813, in <module>
    cost, big_h0, h0 = train_fn(seqs, big_h0, h0, reset, mask)
  File "/usr/local/lib/python2.7/dist-packages/theano/compile/function_module.py", line 898, in __call__
    storage_map=getattr(self.fn, 'storage_map', None))
  File "/usr/local/lib/python2.7/dist-packages/theano/gof/link.py", line 325, in raise_with_op
    reraise(exc_type, exc_value, exc_trace)
  File "/usr/local/lib/python2.7/dist-packages/theano/compile/function_module.py", line 884, in __call__
    self.fn() if output_subset is None else\
IndexError: One of the index value is out of bound. Error code: 65535.\n
Apply node that caused the error: GpuAdvancedSubtensor1(SampleLevel.Embedding, Elemwise{Cast{int64}}.0)
Toposort index: 283
Inputs types: [CudaNdarrayType(float32, matrix), TensorType(int64, vector)]
Inputs shapes: [(256, 256), (131072,)]
Inputs strides: [(256, 1), (8,)]
Inputs values: ['not shown', 'not shown']
Outputs clients: [[GpuReshape{2}(GpuAdvancedSubtensor1.0, TensorConstant{[ -1 512]})]]

HINT: Re-running with most Theano optimization disabled could give you a back-trace of when this node was created. This can be done with by setting the Theano flag 'optimizer=fast_compile'. If that does not work, Theano optimizations can be disabled with 'optimizer=None'.
HINT: Use the Theano flag 'exception_verbosity=high' for a debugprint and storage map footprint of this apply node.
```

**Comment by greaber** [comment]
Maybe this is just reading tea leaves, but without weight norm, seemingly the network was already somewhat overtrained (although worse on validation than test for some reason):
```
/usr/local/lib/python2.7/dist-packages/theano/gradient.py:539: UserWarning: grad method was asked to compute the gradient with respect to a variable that is not part of the computational graph of the cost, or is used only by a non-differentiable operator: BigFrameLevel.IndependentPreds.W0
  handle_disconnected(elem)
/usr/local/lib/python2.7/dist-packages/theano/gradient.py:539: UserWarning: grad method was asked to compute the gradient with respect to a variable that is not part of the computational graph of the cost, or is used only by a non-differentiable operator: BigFrameLevel.IndependentPreds.b
  handle_disconnected(elem)
/usr/local/lib/python2.7/dist-packages/theano/gradient.py:565: UserWarning: grad method was asked to compute the gradient with respect to a variable that is not part of the computational graph of the cost, or is used only by a non-differentiable operator: <DisconnectedType>
  handle_disconnected(rval[i])
Wall clock time spent before training started: 0.02h
Training!
0

 500 1000 1500 2000 2500 3000 3500 4000
Validation! Done!

>>> Best validation cost of 2.29261279106 reached. Testing! Done!
>>> test cost:2.16458010674     total time:138.77671504
epoch:0 total iters:4082        wall clock time:1.60h
>>> Lowest valid cost:2.29261279106      Corresponding test cost:2.16458010674
        train cost:2.0642       total time:1.50h        per iter:1.323s
        valid cost:2.2926       total time:0.04h
        test  cost:2.1646       total time:0.04h
Saving params! Done!
Train info saved! And plotted!
Sampling! 20 samples of 5 seconds length generated in 258.886137962 seconds. Done!
Validation Done!
Back to Training...
4500 5000 5500 6000
```
But with weight norm, strangely the training cost was higher than validation and test:

```
/usr/local/lib/python2.7/dist-packages/theano/gradient.py:539: UserWarning: grad method was asked to compute the gradient with respect to a variable that is not par
t of the computational graph of the cost, or is used only by a non-differentiable operator: BigFrameLevel.IndependentPreds.W0
  handle_disconnected(elem)
/usr/local/lib/python2.7/dist-packages/theano/gradient.py:539: UserWarning: grad method was asked to compute the gradient with respect to a variable that is not par
t of the computational graph of the cost, or is used only by a non-differentiable operator: BigFrameLevel.IndependentPreds.g0
  handle_disconnected(elem)
/usr/local/lib/python2.7/dist-packages/theano/gradient.py:539: UserWarning: grad method was asked to compute the gradient with respect to a variable that is not par
t of the computational graph of the cost, or is used only by a non-differentiable operator: BigFrameLevel.IndependentPreds.b
  handle_disconnected(elem)
/usr/local/lib/python2.7/dist-packages/theano/gradient.py:565: UserWarning: grad method was asked to compute the gradient with respect to a variable that is not par
t of the computational graph of the cost, or is used only by a non-differentiable operator: <DisconnectedType>
  handle_disconnected(rval[i])
Wall clock time spent before training started: 0.03h
Training!
0 500 1000 1500 2000 2500 3000 3500
Validation! Done!

>>> Best validation cost of 1.59446287155 reached. Testing! Done!
>>> test cost:1.48137795925     total time:134.307480812
epoch:0 total iters:3669        wall clock time:1.60h
>>> Lowest valid cost:1.59446287155      Corresponding test cost:1.48137795925
        train cost:1.6993       total time:1.50h        per iter:1.472s
        valid cost:1.5945       total time:0.04h
        test  cost:1.4814       total time:0.04h
Saving params! Done!
Train info saved! And plotted!
Sampling! 20 samples of 5 seconds length generated in 298.524422884 seconds. Done!
Validation Done!
Back to Training...
4000 4500 5000 5500 6000
```

**Comment by greaber** [comment]
Probably just random variation because it bombed out so early I guess...

**Comment by richardassar** [comment]
I typed a reply but the gmail mobile app still has an annoying bug that I
experienced years ago where it loses your message after hitting send.
Thanks Google.

Anyway...

Reducing the learning rate did not fix the problem for me, neither did
reducing the gradient clipping threshold - which is a per-dimension
clipping to a hypercube, a very common approach first suggested by Alex
Graves iirc. This means the gradient can still be very large.

I tried clipping the grad norm instead but the problem persisted. You
either don't converge or if you do you get the huge gradient spikes.

The per-parameter statistics that Adam collects might contribute to the
problem when it arises. Normally the gradient jitters around but there are
a few iterations where the gradient norm increases slightly before
exploding. Sometimes it does so on the first TBPTT step, other times later.
There are signs it's about to happen way before it does.

Reducing the number of TBPTT steps, switching to SGD without momentum or
reducing the batch size will just slow training down. They provide no
garauntee that the problem has been addressed.

Weight normalisation and layer normalisation provide a mechanism to
stabilise the hidden state dynamics of RNNs, I imagine both work just as
effectively for SampleRNN but it would be interesting to compare LN vs WN.

In section 1 of the Weightnorm paper there are a few statements concerning
cost gradient whitening and the condition number of the Hessian which
justify reparameterisation.

On 2 Jun 2017 22:01, "greaber" <notifications@github.com> wrote:

OK, I am running the reference code now, and I will let you know how it
goes.

So your guess is that there is a single bad update step that kills the
network, and that the reason it is possible to have such a step is that
with TBPTT you can get larger steps (even though we clip the gradient)? And
using an optimizer like Adam or rmsprop that tries to do bigger updates of
rarely updated parameters can make it worse because it might be that it is
precisely seldom-updated parameters that cause the problem? However, there
is no numerical problem, it is just that we are getting into a pathological
region of parameter space? I guess this sounds plausible, but I'm not sure
if it is true.

I have experienced a similar phenomenon (but possibly it had a different
origin) when training a Wavenet, so I wonder if TBPTT is critical or not.
In the wavenet case, at least sometimes, reducing the learning rate was
enough to fix it. I never tried weight norm with the wavenet, but it could
be a good idea.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/richardassar/SampleRNN_torch/issues/1#issuecomment-305779503>,
or mute the thread
